File: computing/interview/data-structures/python/trie.py
# Trie (Prefix Tree) — O(m) insert/search/prefix-check (m = key length)


class TrieNode:
    def __init__(self):
        self.children: dict[str, "TrieNode"] = {}
        self.is_end = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str):          # O(m)
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        node.is_end = True

    def search(self, word: str) -> bool:  # O(m) — exact match
        node = self._find(word)
        return node is not None and node.is_end

    def starts_with(self, prefix: str) -> bool:  # O(m)
        return self._find(prefix) is not None

    def autocomplete(self, prefix: str) -> list[str]:
        """Return all words starting with prefix."""
        node = self._find(prefix)
        if not node:
            return []
        results: list[str] = []
        self._collect(node, prefix, results)
        return results

    def _find(self, s: str):
        node = self.root
        for ch in s:
            if ch not in node.children:
                return None
            node = node.children[ch]
        return node

    def _collect(self, node: TrieNode, current: str, results: list[str]):
        if node.is_end:
            results.append(current)
        for ch, child in node.children.items():
            self._collect(child, current + ch, results)


# --- Demo ---
trie = Trie()
for w in ["apple", "app", "application", "apply", "banana"]:
    trie.insert(w)

print("search('app'):", trie.search("app"))           # True
print("search('apple'):", trie.search("apple"))       # True
print("search('ap'):", trie.search("ap"))             # False
print("starts_with('app'):", trie.starts_with("app")) # True
print("starts_with('cat'):", trie.starts_with("cat")) # False
print("autocomplete('app'):", sorted(trie.autocomplete("app")))
print("autocomplete('ban'):", trie.autocomplete("ban"))

File: computing/interview/data-structures/cpp/trie.cpp
// Trie (Prefix Tree) — stores strings for O(m) insert/search/prefix-check
#include <array>
#include <iostream>
#include <string>

struct TrieNode {
    std::array<TrieNode*, 26> children{};
    bool isEnd = false;

    ~TrieNode() {
        for (TrieNode* child : children) delete child;
    }
};

class Trie {
public:
    Trie() : root(new TrieNode()) {}
    ~Trie() { delete root; }

    // O(m) — m = word length
    void insert(const std::string& word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx]) node->children[idx] = new TrieNode();
            node = node->children[idx];
        }
        node->isEnd = true;
    }

    // O(m) — exact word search
    bool search(const std::string& word) const {
        TrieNode* node = find(word);
        return node && node->isEnd;
    }

    // O(m) — check if any word has this prefix
    bool startsWith(const std::string& prefix) const {
        return find(prefix) != nullptr;
    }

private:
    TrieNode* root;

    TrieNode* find(const std::string& s) const {
        TrieNode* node = root;
        for (char c : s) {
            int idx = c - 'a';
            if (!node->children[idx]) return nullptr;
            node = node->children[idx];
        }
        return node;
    }
};

int main() {
    Trie trie;
    for (const std::string& w : {"apple", "app", "application", "apply", "banana"})
        trie.insert(w);

    std::cout << std::boolalpha;
    std::cout << "search(\"app\"):         " << trie.search("app")         << "\n";
    std::cout << "search(\"apple\"):       " << trie.search("apple")       << "\n";
    std::cout << "search(\"ap\"):          " << trie.search("ap")          << "\n";
    std::cout << "startsWith(\"app\"):     " << trie.startsWith("app")     << "\n";
    std::cout << "startsWith(\"ban\"):     " << trie.startsWith("ban")     << "\n";
    std::cout << "startsWith(\"cat\"):     " << trie.startsWith("cat")     << "\n";

    return 0;
}
